helix-term/src/commands/lsp.rs

```diff
@@ -452,43 +452,76 @@ pub fn code_action(cx: &mut Context) {
     cx.callback(
         future,
         move |editor, compositor, response: Option<lsp::CodeActionResponse>| {
-            let actions = match response {
+            let mut actions = match response {
                 Some(a) => a,
                 None => return,
             };
+
             if actions.is_empty() {
                 editor.set_status("No code actions available");
                 return;
             }
 
-            let mut picker = ui::Menu::new(actions, (), move |editor, code_action, event| {
-                if event != PromptEvent::Validate {
-                    return;
-                }
+            // sort by CodeActionKind
+            // this ensures that the most relevant codeactions (quickfix) show up first
+            // while more situational commands (like refactors) show up later
+            // this behaviour is modeled after the behaviour of vscode (editor/contrib/codeAction/browser/codeActionWidget.ts)
+
+            let mut categories = vec![Vec::new(); 8];
+            for action in actions.drain(..) {
+                let category = match &action {
+                    lsp::CodeActionOrCommand::CodeAction(lsp::CodeAction {
+                        kind: Some(kind),
+                        ..
+                    }) => match kind.as_str() {
```

Comment:
Why match on the str form when you could match on `Some(Enum::Member)` and `None` ?

Reply:
Or, here, since you match on `Some(kind)` above, directly on `Enum::Member`

Reply:
@poliorcetics  `CodeActionKind` isn't an Enum its:
```rust
#[derive(Debug, Eq, PartialEq, Hash, PartialOrd, Clone, Deserialize, Serialize)]
pub struct CodeActionKind(Cow<'static, str>);
```
(from `lsp-types` crate)

Although maybe we should match against the consts in `CodeActionKind` rather than on strings directly :
```rust
    pub const EMPTY: CodeActionKind = CodeActionKind::new("");
    pub const QUICKFIX: CodeActionKind = CodeActionKind::new("quickfix");
    pub const REFACTOR: CodeActionKind = CodeActionKind::new("refactor");
    pub const REFACTOR_EXTRACT: CodeActionKind = CodeActionKind::new("refactor.extract");
    pub const REFACTOR_INLINE: CodeActionKind = CodeActionKind::new("refactor.inline");
    pub const REFACTOR_REWRITE: CodeActionKind = CodeActionKind::new("refactor.rewrite");
    pub const SOURCE: CodeActionKind = CodeActionKind::new("source");
    pub const SOURCE_ORGANIZE_IMPORTS: CodeActionKind = CodeActionKind::new("source.organizeImports");
```

Also currently you're treating `source.organizeImports` and `source.fixAll`[^1] as other (7), and not as part of `source` (6).

[^1]: Not present in lsp-types crate but present in spec

Reply:
Matching on the constants does not work because CodeActionKind is a `Cow` which does not implement structural Eq (so can not be used in a match) while the string can be used in the match.

Upon @A-Walrus suggestion I looked more closely at VScode and noticed that they treat the `CodeActionKind` as a path (seperated by a `.`). Therefore when they match `quickfix`, they also match `quickfix.X.Y.Z`. I have updated the PR to do the same by using `str::split('.')` and matching on the components.